src/httpfetch.cpp

```diff
@@ -275,55 +275,60 @@ HTTPFetchOngoing::HTTPFetchOngoing(const HTTPFetchRequest &request_,
 		curl_easy_setopt(curl, CURLOPT_WRITEDATA, &result.data);
 	}
 
+	// Configure the method
+	switch (request.method) {
+	case HTTP_GET:
+		curl_easy_setopt(curl, CURLOPT_HTTPGET, 1);
+		break;
+	case HTTP_POST:
+		curl_easy_setopt(curl, CURLOPT_POST, 1);
+		break;
+	case HTTP_PUT:
+		curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "PUT");
+		break;
+	case HTTP_DELETE:
+		curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "DELETE");
+		break;
+	}
+
 	// Set data from fields or raw_data
 	if (request.multipart) {
+		assert(request.method != HTTP_GET);
 		multipart_mime = curl_mime_init(curl);
 		for (auto &it : request.fields) {
 			curl_mimepart *part = curl_mime_addpart(multipart_mime);
 			curl_mime_name(part, it.first.c_str());
 			curl_mime_data(part, it.second.c_str(), it.second.size());
 		}
 		curl_easy_setopt(curl, CURLOPT_MIMEPOST, multipart_mime);
-	} else {
-		switch (request.method) {
-		case HTTP_GET:
-			curl_easy_setopt(curl, CURLOPT_HTTPGET, 1);
-			break;
-		case HTTP_POST:
-			curl_easy_setopt(curl, CURLOPT_POST, 1);
-			break;
-		case HTTP_PUT:
-			curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "PUT");
-			break;
-		case HTTP_DELETE:
-			curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "DELETE");
-			break;
-		}
-		if (request.method != HTTP_GET) {
-			if (!request.raw_data.empty()) {
-				curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE,
-						request.raw_data.size());
-				curl_easy_setopt(curl, CURLOPT_POSTFIELDS,
-						request.raw_data.c_str());
-			} else if (!request.fields.empty()) {
-				std::string str;
-				for (auto &field : request.fields) {
-					if (!str.empty())
-						str += "&";
-					str += urlencode(field.first);
-					str += "=";
-					str += urlencode(field.second);
-				}
-				curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE,
-						str.size());
-				curl_easy_setopt(curl, CURLOPT_COPYPOSTFIELDS,
-						str.c_str());
+	} else if (request.method != HTTP_GET) {
+		if (!request.raw_data.empty()) {
+			curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE,
+					request.raw_data.size());
+			curl_easy_setopt(curl, CURLOPT_POSTFIELDS,
+					request.raw_data.c_str());
+		} else if (!request.fields.empty()) {
+			std::string str;
+			for (auto &field : request.fields) {
+				if (!str.empty())
+					str += "&";
+				str += urlencode(field.first);
+				str += "=";
+				str += urlencode(field.second);
 			}
+			curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, str.size());
+			curl_easy_setopt(curl, CURLOPT_COPYPOSTFIELDS, str.c_str());
+		} else {
+			// We must set the fields to an empty buffer (not NULL!)
+			// or curl would try to read from stdin.
+			curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, 0);
+			curl_easy_setopt(curl, CURLOPT_POSTFIELDS, "");
```

Comment:
Couldn't we merge this with the `if (!request.raw_data.empty()) {` branch above? It would do the same thing.

Reply:
Will do